# What We Can Do

We can get 4 year of financial data cleanly outputted

Link: https://towardsdatascience.com/web-scraping-for-accounting-analysis-using-python-part-1-b5fc016a1c9a


Goal is to be able to quickly analyze data to narrowdown investment opportunities

In [298]:
import pandas as pd
from bs4 import BeautifulSoup
import urllib.request as ur

In [299]:
# Enter a stock symbol
index= "GRUB"
# URL link 
url_is = "https://finance.yahoo.com/quote/" + index + "/financials?p=" + index
url_bs = "https://finance.yahoo.com/quote/" + index +"/balance-sheet?p=" + index
url_cf = "https://finance.yahoo.com/quote/" + index + "/cash-flow?p="+ index


In [300]:
# read the income statement
read_data = ur.urlopen(url_is).read() 
soup_is= BeautifulSoup(read_data,"lxml")

# read in the balance sheet
read_data_bs = ur.urlopen(url_bs).read() 
soup_bs= BeautifulSoup(read_data_bs,"lxml")

# read in the statement of cash flows
read_data_cf = ur.urlopen(url_cf).read() 
soup_cf= BeautifulSoup(read_data_cf,"lxml")

In [301]:
# Creating the list for the income statement
ls= [] # Create empty list
for l in soup_is.find_all("div"): 
  #Find all data structure that is ‘div’
  ls.append(l.string) # add each element one by one to the list
 
ls = [e for e in ls if e not in ("Operating Expenses","Non-recurring Events")] # Exclude those columns


# Creating the list for the balance sheet
ls_bs= [] # Create empty list
for l in soup_bs.find_all("div"): 
  #Find all data structure that is ‘div’
  ls_bs.append(l.string) # add each element one by one to the list
    
# Creating the list for the cash flows
ls_cf= [] # Create empty list
for l in soup_cf.find_all("div"): 
  #Find all data structure that is ‘div’
  ls_cf.append(l.string) # add each element one by one to the list

In [302]:
# income statement filter
new_ls = list(filter(None,ls))

# Balance sheet filter
new_ls_bs = list(filter(None,ls_bs))

# cash flows filter
new_ls_cf = list(filter(None,ls_cf))




In [303]:
# income statemet getting rid of irelavent rows. Make sure to run all above ever time run this
new_ls = new_ls[12:]

# income statemet getting rid of irelavent rows. Make sure to run all above ever time run this
new_ls_bs = new_ls_bs[12:]

# income statemet getting rid of irelavent rows. Make sure to run all above ever time run this
new_ls_cf = new_ls_cf[12:]


In [304]:
# zipping the date for income statement
is_data = list(zip(*[iter(new_ls)]*6))

# zipping the date for BS
bs_data = list(zip(*[iter(new_ls_bs)]*5))

# zipping the date for cash Flow
cf_data = list(zip(*[iter(new_ls_cf)]*6))

In [305]:
Income_st = pd.DataFrame(is_data[0:])

Balance_sheet = pd.DataFrame(bs_data[0:])


Cash_flows = pd.DataFrame(cf_data[0:])


In [306]:
# Cleaning up the Income Statement
Income_st.columns = Income_st.iloc[0] # Name columns to first row of dataframe
Income_st = Income_st.iloc[1:,] # start to read 1st row
# setting the index of each row
Income_st.index=Income_st["Quarterly"].tolist()
Income_st2 = Income_st.drop(["Quarterly", "ttm"], axis=1)

In [307]:
# Cleaning the Balance sheet Data
Balance_sheet.columns = Balance_sheet.iloc[0] # Name columns to first row of dataframe
Balance_sheet = Balance_sheet.iloc[1:,] # start to read 1st row
# setting the index of each row
Balance_sheet.index=Balance_sheet["Quarterly"].tolist()
Balance_sheet2 = Balance_sheet.drop(["Quarterly"], axis=1)

In [308]:
# Cleaning the Balance sheet Data
Cash_flows.columns = Cash_flows.iloc[0] # Name columns to first row of dataframe
Cash_flows = Cash_flows.iloc[1:,] # start to read 1st row
# setting the index of each row
Cash_flows.index=Cash_flows["Quarterly"].tolist()
Cash_flows= Cash_flows.drop(["Quarterly", "ttm"], axis=1)

In [309]:
finalIS = Income_st2
finalBS = Balance_sheet2
finalCF = Cash_flows

In [310]:
finalIS.drop(finalIS.tail(5).index,inplace=True)

In [311]:
finalIS

0                                           12/31/2019 12/31/2018 12/31/2017  \
Total Revenue                                1,312,151  1,007,257    683,067   
Cost of Revenue                                790,768    536,599    325,716   
Gross Profit                                   521,383    470,658    357,351   
Selling General and Administrative             412,217    299,755    215,753   
Total Operating Expenses                       527,666    385,695    267,601   
Operating Income or Loss                        -6,283     84,963     89,750   
Income Before Tax                              -26,776     81,433     89,648   
Income Tax Expense                              -8,210      2,952     -9,335   
Income from Continuing Operations              -18,566     78,481     98,983   
Net Income                                     -18,566     78,481     98,983   
Net Income available to common shareholders    -18,566     78,481     98,983   

0                                           12/31/2016  
Total Revenue                                  493,331  
Cost of Revenue                                214,210  
Gross Profit                                   279,121  
Selling General and Administrative             160,076  
Total Operating Expenses                       195,269  
Operating Income or Loss                        83,852  
Income Before Tax                               83,852  
Income Tax Expense                              34,295  
Income from Continuing Operations               49,557  
Net Income                                      49,557  
Net Income available to common shareholders     49,557

In [312]:
finalBS

0                                          12/31/2019 12/31/2018 12/31/2017  \
Cash And Cash Equivalents                     375,909    211,245    234,090   
Other Short Term Investments                   49,275     14,084     23,605   
Total Cash                                    425,184    225,329    257,695   
Net Receivables                               119,658    110,855     95,970   
Total Current Assets                          566,317    363,775    360,483   
Gross property, plant and equipment           416,178    198,092    116,780   
Accumulated Depreciation                     -142,802    -78,597    -45,396   
Net property, plant and equipment             273,376    119,495     71,384   
Goodwill                                    1,007,968  1,019,239    589,862   
Intangible Assets                             500,481    549,013    515,553   
Other long-term assets                         26,836     14,186      6,487   
Total non-current assets                    1,808,661  1,701,933  1,183,286   
Total Assets                                2,374,978  2,065,708  1,543,769   
Current Debt                                        -      6,250      3,906   
Accounts Payable                               26,748     26,656      7,607   
Accrued liabilities                            69,274     62,496     40,004   
Deferred revenues                              12,212          -          -   
Other Current Liabilities                     131,753    127,344    119,922   
Total Current Liabilities                     249,363    223,168    174,548   
Long Term Debt                                493,009    335,548    169,645   
Deferred taxes liabilities                     27,163     46,383     74,292   
Total non-current liabilities                 632,045    400,201    251,405   
Total Liabilities                             881,408    623,369    425,953   
Common Stock                                        9          9          9   
Retained Earnings                             330,789    349,355    269,992   
Accumulated other comprehensive income         -1,628     -1,891     -1,228   
Total stockholders' equity                  1,493,570  1,442,339  1,117,816   
Total liabilities and stockholders' equity  2,374,978  2,065,708  1,543,769   

0                                          12/31/2016  
Cash And Cash Equivalents                     239,528  
Other Short Term Investments                   84,091  
Total Cash                                    323,619  
Net Receivables                                60,550  
Total Current Assets                          396,337  
Gross property, plant and equipment            69,248  
Accumulated Depreciation                      -22,693  
Net property, plant and equipment              46,555  
Goodwill                                      436,455  
Intangible Assets                             313,630  
Other long-term assets                          4,530  
Total non-current assets                      801,170  
Total Assets                                1,197,507  
Current Debt                                        -  
Accounts Payable                                7,590  
Accrued liabilities                            18,686  
Deferred revenues                                   -  
Other Current Liabilities                      83,349  
Total Current Liabilities                     110,490  
Long Term Debt                                      -  
Deferred taxes liabilities                    108,022  
Total non-current liabilities                 114,898  
Total Liabilities                             225,388  
Common Stock                                        9  
Retained Earnings                             168,457  
Accumulated other comprehensive income         -2,078  
Total stockholders' equity                    972,119  
Total liabilities and stockholders' equity  1,197,507

# Financial Statement Analysis

In [313]:
# Converting all strings to numbers in balance sheet
# Remove all the commas, and change all - to 0
bsAnalysis = finalBS

# replaces commas with nothing
bsAnalysis["12/31/2019"] = bsAnalysis["12/31/2019"].str.replace(',','')
# be able to replace - with 0 
bsAnalysis["12/31/2019"] = bsAnalysis["12/31/2019"].str.replace('-','0')
# changes type to int
bsAnalysis["12/31/2019"] = bsAnalysis["12/31/2019"].astype(int)


# replaces commas with nothing
bsAnalysis["12/31/2018"] = bsAnalysis["12/31/2018"].str.replace(',','')
# be able to replace - with 0 
bsAnalysis["12/31/2018"] = bsAnalysis["12/31/2018"].str.replace('-','0')
# changes type to int
bsAnalysis["12/31/2018"] = bsAnalysis["12/31/2018"].astype(int)

# replaces commas with nothing
bsAnalysis["12/31/2017"] = bsAnalysis["12/31/2017"].str.replace(',','')
# be able to replace - with 0 
bsAnalysis["12/31/2017"] = bsAnalysis["12/31/2017"].str.replace('-','0')
# changes type to int
bsAnalysis["12/31/2017"] = bsAnalysis["12/31/2017"].astype(int)


# replaces commas with nothing
bsAnalysis["12/31/2016"] = bsAnalysis["12/31/2016"].str.replace(',','')
# be able to replace - with 0 
bsAnalysis["12/31/2016"] = bsAnalysis["12/31/2016"].str.replace('-','0')
# changes type to int
bsAnalysis["12/31/2016"] = bsAnalysis["12/31/2016"].astype(int)

In [314]:
bsAnalysis

0                                           12/31/2019  12/31/2018  \
Cash And Cash Equivalents                       375909      211245   
Other Short Term Investments                     49275       14084   
Total Cash                                      425184      225329   
Net Receivables                                 119658      110855   
Total Current Assets                            566317      363775   
Gross property, plant and equipment             416178      198092   
Accumulated Depreciation                        142802       78597   
Net property, plant and equipment               273376      119495   
Goodwill                                       1007968     1019239   
Intangible Assets                               500481      549013   
Other long-term assets                           26836       14186   
Total non-current assets                       1808661     1701933   
Total Assets                                   2374978     2065708   
Current Debt                                         0        6250   
Accounts Payable                                 26748       26656   
Accrued liabilities                              69274       62496   
Deferred revenues                                12212           0   
Other Current Liabilities                       131753      127344   
Total Current Liabilities                       249363      223168   
Long Term Debt                                  493009      335548   
Deferred taxes liabilities                       27163       46383   
Total non-current liabilities                   632045      400201   
Total Liabilities                               881408      623369   
Common Stock                                         9           9   
Retained Earnings                               330789      349355   
Accumulated other comprehensive income            1628        1891   
Total stockholders' equity                     1493570     1442339   
Total liabilities and stockholders' equity     2374978     2065708   

0                                           12/31/2017  12/31/2016  
Cash And Cash Equivalents                       234090      239528  
Other Short Term Investments                     23605       84091  
Total Cash                                      257695      323619  
Net Receivables                                  95970       60550  
Total Current Assets                            360483      396337  
Gross property, plant and equipment             116780       69248  
Accumulated Depreciation                         45396       22693  
Net property, plant and equipment                71384       46555  
Goodwill                                        589862      436455  
Intangible Assets                               515553      313630  
Other long-term assets                            6487        4530  
Total non-current assets                       1183286      801170  
Total Assets                                   1543769     1197507  
Current Debt                                      3906           0  
Accounts Payable                                  7607        7590  
Accrued liabilities                              40004       18686  
Deferred revenues                                    0           0  
Other Current Liabilities                       119922       83349  
Total Current Liabilities                       174548      110490  
Long Term Debt                                  169645           0  
Deferred taxes liabilities                       74292      108022  
Total non-current liabilities                   251405      114898  
Total Liabilities                               425953      225388  
Common Stock                                         9           9  
Retained Earnings                               269992      168457  
Accumulated other comprehensive income            1228        2078  
Total stockholders' equity                     1117816      972119  
Total liabilities and stockholders' e

In [315]:
# selecting specific values
bsAnalysis.loc["Total Cash", "12/31/2019" ]

425184

In [316]:
bsAnalysis

0                                           12/31/2019  12/31/2018  \
Cash And Cash Equivalents                       375909      211245   
Other Short Term Investments                     49275       14084   
Total Cash                                      425184      225329   
Net Receivables                                 119658      110855   
Total Current Assets                            566317      363775   
Gross property, plant and equipment             416178      198092   
Accumulated Depreciation                        142802       78597   
Net property, plant and equipment               273376      119495   
Goodwill                                       1007968     1019239   
Intangible Assets                               500481      549013   
Other long-term assets                           26836       14186   
Total non-current assets                       1808661     1701933   
Total Assets                                   2374978     2065708   
Current Debt                                         0        6250   
Accounts Payable                                 26748       26656   
Accrued liabilities                              69274       62496   
Deferred revenues                                12212           0   
Other Current Liabilities                       131753      127344   
Total Current Liabilities                       249363      223168   
Long Term Debt                                  493009      335548   
Deferred taxes liabilities                       27163       46383   
Total non-current liabilities                   632045      400201   
Total Liabilities                               881408      623369   
Common Stock                                         9           9   
Retained Earnings                               330789      349355   
Accumulated other comprehensive income            1628        1891   
Total stockholders' equity                     1493570     1442339   
Total liabilities and stockholders' equity     2374978     2065708   

0                                           12/31/2017  12/31/2016  
Cash And Cash Equivalents                       234090      239528  
Other Short Term Investments                     23605       84091  
Total Cash                                      257695      323619  
Net Receivables                                  95970       60550  
Total Current Assets                            360483      396337  
Gross property, plant and equipment             116780       69248  
Accumulated Depreciation                         45396       22693  
Net property, plant and equipment                71384       46555  
Goodwill                                        589862      436455  
Intangible Assets                               515553      313630  
Other long-term assets                            6487        4530  
Total non-current assets                       1183286      801170  
Total Assets                                   1543769     1197507  
Current Debt                                      3906           0  
Accounts Payable                                  7607        7590  
Accrued liabilities                              40004       18686  
Deferred revenues                                    0           0  
Other Current Liabilities                       119922       83349  
Total Current Liabilities                       174548      110490  
Long Term Debt                                  169645           0  
Deferred taxes liabilities                       74292      108022  
Total non-current liabilities                   251405      114898  
Total Liabilities                               425953      225388  
Common Stock                                         9           9  
Retained Earnings                               269992      168457  
Accumulated other comprehensive income            1228        2078  
Total stockholders' equity                     1117816      972119  
Total liabilities and stockholders' e

In [317]:
cfAnalysis = finalCF

# replaces commas with nothing
cfAnalysis["12/31/2019"] = cfAnalysis["12/31/2019"].str.replace(',','')
# be able to replace - with 0 
cfAnalysis["12/31/2019"] = cfAnalysis["12/31/2019"].str.replace('-','0')
# changes type to int
cfAnalysis["12/31/2019"] = cfAnalysis["12/31/2019"].astype(int)


# replaces commas with nothing
cfAnalysis["12/31/2018"] = cfAnalysis["12/31/2018"].str.replace(',','')
# be able to replace - with 0 
cfAnalysis["12/31/2018"] = cfAnalysis["12/31/2018"].str.replace('-','0')
# changes type to int
cfAnalysis["12/31/2018"] = cfAnalysis["12/31/2018"].astype(int)

# replaces commas with nothing
cfAnalysis["12/31/2017"] = cfAnalysis["12/31/2017"].str.replace(',','')
# be able to replace - with 0 
cfAnalysis["12/31/2017"] = cfAnalysis["12/31/2017"].str.replace('-','0')
# changes type to int
cfAnalysis["12/31/2017"] = cfAnalysis["12/31/2017"].astype(int)


# replaces commas with nothing
cfAnalysis["12/31/2016"] = cfAnalysis["12/31/2016"].str.replace(',','')
# be able to replace - with 0 
cfAnalysis["12/31/2016"] = cfAnalysis["12/31/2016"].str.replace('-','0')
# changes type to int
cfAnalysis["12/31/2016"] = cfAnalysis["12/31/2016"].astype(int)



In [318]:
cfAnalysis

0                                                   12/31/2019  12/31/2018  \
Net Income                                               18566       78481   
Depreciation & amortization                             115449       85940   
Deferred income taxes                                     7726        1724   
Stock based compensation                                 72879       55261   
Change in working capital                                12055        1431   
Accounts receivable                                      11591        6092   
Accounts Payable                                          1978       11160   
Other working capital                                    68951      139463   
Other non-cash items                                      8531        5448   
Net cash provided by operating activites                182622      225527   
Investments in property, plant and equipment            113671       86064   
Acquisitions, net                                            0      517909   
Purchases of investments                                 85989       57197   
Sales/Maturities of investments                          51366       67166   
Other investing activites                                  250           0   
Net cash used for investing activites                   148417      594004   
Debt repayment                                          342313       53906   
Common stock issued                                          0      200000   
Common stock repurchased                                     0           0   
Other financing activites                                32889       35599   
Net cash used privided by (used for) financing ...      129267      346685   
Net change in cash                                      163472       21792   
Cash at beginning of period                             215802      238239   
Cash at end of period                                   379594      215802   
Operating Cash Flow                                     182622      225527   
Capital Expenditure                                     113671       86064   
Free Cash Flow                                           68951      139463   

0                                                   12/31/2017  12/31/2016  
Net Income                                               98983       49557  
Depreciation & amortization                              51848       35193  
Deferred income taxes                                    31179        1027  
Stock based compensation                                 32748       23559  
Change in working capital                                 2117       13640  
Accounts receivable                                      27833       17488  
Accounts Payable                                          4244        3204  
Other working capital                                    87297       60532  
Other non-cash items                                      1772         880  
Net cash provided by operating activites                152740       97678  
Investments in property, plant and equipment             65443       37146  
Acquisitions, net                                       333301       65849  
Purchases of investments                                154758      226694  
Sales/Maturities of investments                         215983      284662  
Other investing activites                                  557         492  
Net cash used for investing activites                   336962       45519  
Debt repayment                                           25781           0  
Common stock issued                                          0           0  
Common stock repurchased                                     0       14774  
Other financing activites                                12535       20650  
Net cash used privided by (used for) financing ...      178059       19344  
Net change in cash                                        6163       71503  
Cash at beginning of period                 

In [319]:
isAnalysis = finalIS

# replaces commas with nothing
isAnalysis["12/31/2019"] = isAnalysis["12/31/2019"].str.replace(',','')
# be able to replace - with 0 
isAnalysis["12/31/2019"] = isAnalysis["12/31/2019"].str.replace('-','0')
# changes type to int
isAnalysis["12/31/2019"] = isAnalysis["12/31/2019"].astype(int)


# replaces commas with nothing
isAnalysis["12/31/2018"] = isAnalysis["12/31/2018"].str.replace(',','')
# be able to replace - with 0 
isAnalysis["12/31/2018"] = isAnalysis["12/31/2018"].str.replace('-','0')
# changes type to int
isAnalysis["12/31/2018"] = isAnalysis["12/31/2018"].astype(int)

# replaces commas with nothing
isAnalysis["12/31/2017"] = isAnalysis["12/31/2017"].str.replace(',','')
# be able to replace - with 0 
isAnalysis["12/31/2017"] = isAnalysis["12/31/2017"].str.replace('-','0')
# changes type to int
isAnalysis["12/31/2017"] = isAnalysis["12/31/2017"].astype(int)


# replaces commas with nothing
isAnalysis["12/31/2016"] = isAnalysis["12/31/2016"].str.replace(',','')
# be able to replace - with 0 
isAnalysis["12/31/2016"] = isAnalysis["12/31/2016"].str.replace('-','0')
# changes type to int
isAnalysis["12/31/2016"] = isAnalysis["12/31/2016"].astype(int)

In [320]:
isAnalysis

0                                            12/31/2019  12/31/2018  \
Total Revenue                                   1312151     1007257   
Cost of Revenue                                  790768      536599   
Gross Profit                                     521383      470658   
Selling General and Administrative               412217      299755   
Total Operating Expenses                         527666      385695   
Operating Income or Loss                           6283       84963   
Income Before Tax                                 26776       81433   
Income Tax Expense                                 8210        2952   
Income from Continuing Operations                 18566       78481   
Net Income                                        18566       78481   
Net Income available to common shareholders       18566       78481   

0                                            12/31/2017  12/31/2016  
Total Revenue                                    683067      493331  
Cost of Revenue                                  325716      214210  
Gross Profit                                     357351      279121  
Selling General and Administrative               215753      160076  
Total Operating Expenses                         267601      195269  
Operating Income or Loss                          89750       83852  
Income Before Tax                                 89648       83852  
Income Tax Expense                                 9335       34295  
Income from Continuing Operations                 98983       49557  
Net Income                                        98983       49557  
Net Income available to common shareholders       98983       49557

In [321]:
# selecting specific values
isAnalysis.loc["Total Revenue", "12/31/2019" ]

1312151

# 1) Viewing the Cleaned Data

isAnalysis to view the income statement
bsAnalysis to view the balance sheet
cfAnalysis to view the full cash flows


# 2) Getting specific values from the financial statements

EX: isAnalysis.loc["Total Revenue", "12/31/2019"] will get you 1312151


Calculations are simple

GP Margin = isAnalyis.loc["Gross Profit", "12/31/2019"]/isAnalysis.loc["Total Revenue", "12/31/2019"]



In [325]:
# Example of gross margin calculations
GPMargin = isAnalysis.loc["Gross Profit", "12/31/2019"]/isAnalysis.loc["Total Revenue", "12/31/2019"]
GPMargin


0.39734984769283416